### PR TITLE
Admin teacher_pending modification

### PR DESCRIPTION
**Pull request by hannah-whelan**
https://github.com/isaaccomputerscience/isaac-cs-issues/issues/244

### CODE REVIEW

File: config-templates/content_indices.properties [hunk 1]
```diff
@@ -1,4 +1,4 @@
 #
 #Tue Jun 18 15:45:09 BST 2019
-latest=849c7197c96625b651c6cab221ffe652d2d6765c
-live=849c7197c96625b651c6cab221ffe652d2d6765c
+latest=05bbbb74fd52cd0aea3a048a3c1e6c434ce5efaf
```
Comment (jmgq) [inline config-templates/content_indices.properties:3]: Are these changes actually necessary? The file is inside a folder called `config-templates`, so I would imagine that these are just examples that you copy somewhere else, and then update with the right values.

If they are truly necessary, then the comment above should also be updated, as I'd imagine that that's not the right date and time anymore.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/GroupManagerTest.java [hunk 2]
```diff
@@ -130,7 +132,7 @@ final void groupManager_createValidGroup_aGroupShouldBeCreated() {
       assertEquals(someGroupName, capturedGroup.getValue().getGroupName());
       assertInstanceOf(Date.class, capturedGroup.getValue().getCreated());
 
-    } catch (SegueDatabaseException e) {
+    } catch (SegueDatabaseException | NoUserException e) {
       fail("No exception expected");
       e.printStackTrace();
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/GroupManagerTest.java:137]: Not really your fault, but since you're changing code in this area, could you remove this line and instead just pass `e` as the second parameter of `fail`, please?

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 3]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+   *
+   * @param request - to help determine access rights.
+   * @param status    - new teacher pending status.
+   * @param userIds - a list of user ids to change en-mass
+   * @return Success shown by returning an ok response
+   */
+  @POST
+  @Path("/users/teacher_pending/{status}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:308]: This is missing the Swagger documentation (see the `@Operation` annotation in this file for some examples).

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 4]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+  /**
+   * This method will allow users to have their teacher pending status mass changed.
+   *
+   * @param request - to help determine access rights.
+   * @param status    - new teacher pending status.
+   * @param userIds - a list of user ids to change en-mass
+   * @return Success shown by returning an ok response
+   */
+  @POST
+  @Path("/users/teacher_pending/{status}")
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:306]: There are other endpoints under `/admin/` with names such as:
- `/admin/users/change_email_verification_status/{emailVerificationStatus}/{checkEmailsExistBeforeApplying}`
- `/admin/users/change_role/{role}`

They're not great names, but perhaps we can rename this one so that it's in line with those. Something like `/users/change_teacher_pending/{status}`.

None of the options are very RESTFul, unfortunately.

What do you think?

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 5]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+        }
+
+        if (failedUpdates.containsKey(FAILED_TO_SEND)) {
+          errorMessage += String.format("Emails could not be sent to userIds: %s",
+              failedUpdates.get(FAILED_TO_SEND));
+        }
+        return new SegueErrorResponse(Status.BAD_REQUEST, errorMessage).toResponse();
+      }
+
+    } catch (NoUserLoggedInException e) {
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:341]: You're surrounding the whole method between `try`, when in reality only a handful of lines need to be surrounded by it. Given that all you're doing when caching these exceptions is to return an error response, it would be better to only surround the parts of the code that can cause these exceptions.

In the case of `NoUserLoggedInException`, I think you only need it in the first block:

```java
      RegisteredUserDTO requestingUser = userManager.getCurrentRegisteredUser(request);
      if (!isUserAnAdminOrEventManager(userManager, requestingUser)) {
        return new SegueErrorResponse(Status.FORBIDDEN, ACCESS_DENIED_MESSAGE)
            .toResponse();
      }
```

In the case of `SegueDatabaseException`, it only happens in the `for` loop.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 6]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+        if (failedUpdates.containsKey(USERS_NOT_FOUND)) {
+          errorMessage = String.format("One or more users could not be found: %s. ",
+              failedUpdates.get(USERS_NOT_FOUND));
+        }
+
+        if (failedUpdates.containsKey(FAILED_TO_SEND)) {
+          errorMessage += String.format("Emails could not be sent to userIds: %s",
+              failedUpdates.get(FAILED_TO_SEND));
+        }
+        return new SegueErrorResponse(Status.BAD_REQUEST, errorMessage).toResponse();
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:338]: A user not found could be considered a bad request, yes, but the other option (failed to send email) sounds like a server error, not a user error.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 7]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+      }
+
+    } catch (NoUserLoggedInException e) {
+      return SegueErrorResponse.getNotLoggedInResponse();
+    } catch (SegueDatabaseException e) {
+      log.error("Database error while trying to change teacher_pending status", e);
+      return new SegueErrorResponse(Status.INTERNAL_SERVER_ERROR,
+          "Could not save new teacher_pending status to the database").toResponse();
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:346]: Given that this is the message that the user would see (I believe), then I wouldn't mention anything about databases. That's good for the logs, but for the end user all they need to know is that the `teacher_pending` status could not be updated.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 8]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+    return Response.ok().build();
+  }
+
+  private void modifyTeacherPendingStatusForUser(Long userId, RegisteredUserDTO requestingUser, Boolean status,
+                                                 Map<String, List<Long>> failedUpdates) throws SegueDatabaseException {
+
+    try {
+      RegisteredUserDTO user = this.userManager.getUserDTOById(userId);
+
+      if (null == user) {
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:358]: Can this actually happen? After a quick look at the code, it doesn't seem to be an actual possibility.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 9]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+        throw new NoUserException("No user found with this ID.");
+      }
+
+      Boolean oldStatus = user.getTeacherPending();
+      this.userManager.updateTeacherPendingFlag(userId, status);
+      if (log.isInfoEnabled()) {
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:364]: I don't think this is needed. You can simply call `log.info`, and if it's not enabled then it would be a no-op.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 10]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+        log.info("ADMIN user {} has modified the teacher_pending status of {} [{}] from {} to {}",
+            requestingUser.getEmail(), user.getEmail(), user.getId(), oldStatus, status);
+      }
+      if (Boolean.FALSE.equals(status)) {
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:368]: Why is this being compared in this way instead of `if (!status) {`? Is it because we can send `null` values? Is that even allowed? If it's not allowed, perhaps the type needs changing from `Boolean` to `boolean`.

File: config-templates/content_indices.properties (hunk 1 above)
Comment (hannah-whelan) [reply config-templates/content_indices.properties:3]: I think it's a good idea to update this here, as the new endpoint uses a piece of content introduced in this SHA. If we leave it on the old SHA in the template then it essentially sets it up to fail (locally at least) for any dev who doesn't think to update it.  I will change the comment to reflect the date of the SHA

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 4 above)
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:306]: Yeah that makes sense!

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 6 above)
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:338]: Essentially what I was trying to achieve was that it wouldn't cause the front end to change into a server error page if an email failed to send so I was trying to avoid a 500 error (also only wanted it to give an error after having gone through the for loop of users, not erroring out part way through). If the error is under 500 the user (in this case an ADMIN user) gets a popup on the front end to inform them of an error and any message details.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 11]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+        List<Long> userList = failedUpdates.get(USERS_NOT_FOUND);
+        userList.add(userId);
+      }
+    }
+  }
+
+  private void sendTeacherDeclineEmail(RegisteredUserDTO user, Map<String, List<Long>> failedUpdates) {
+    try {
+      emailManager.sendTemplatedEmailToUser(user, emailManager.getEmailTemplateDTO("teacher_declined"),
+          Collections.<String, Object>emptyMap(), EmailType.SYSTEM);
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:387]: You don't need to specify `<String, Object>` here, you can remove that.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 8 above)
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:358]: I'm not sure on this one; I mimicked what was done in `modifyUsersRole`

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 12]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+      log.error("Database error while trying to change teacher_pending status", e);
+      return new SegueErrorResponse(Status.INTERNAL_SERVER_ERROR,
+          "Could not save new teacher_pending status to the database").toResponse();
+    }
+
+    return Response.ok().build();
+  }
+
+  private void modifyTeacherPendingStatusForUser(Long userId, RegisteredUserDTO requestingUser, Boolean status,
+                                                 Map<String, List<Long>> failedUpdates) throws SegueDatabaseException {
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:353]: How come you're throwing this exception instead of catching it and logging it as an error? Is it because you want the whole operation to stop as soon as you encounter a DB error? A consequence of that is that if you have 10 users to update and the exception is thrown while processing user number 3, then 2 users would have been updated and 8 wouldn't. Have you thought about the implications of that scenario? I'm not saying that what you've done is wrong, just mentioning a possible scenario.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 13]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+        userList.add(userId);
+        failedUpdates.put(USERS_NOT_FOUND, userList);
+      } else {
+        List<Long> userList = failedUpdates.get(USERS_NOT_FOUND);
+        userList.add(userId);
+      }
+    }
+  }
+
+  private void sendTeacherDeclineEmail(RegisteredUserDTO user, Map<String, List<Long>> failedUpdates) {
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:384]: I think this was meant to be Decline**d**.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+import uk.ac.cam.cl.dtg.segue.comm.EmailType;
+import uk.ac.cam.cl.dtg.segue.dao.ILogManager;
+import uk.ac.cam.cl.dtg.segue.dao.SegueDatabaseException;
+import uk.ac.cam.cl.dtg.segue.dao.content.ContentManagerException;
+import uk.ac.cam.cl.dtg.segue.dao.content.GitContentManager;
+import uk.ac.cam.cl.dtg.segue.scheduler.SegueJobService;
+import uk.ac.cam.cl.dtg.util.PropertiesLoader;
+
+class AdminFacadeTest {
+  private PropertiesLoader properties;
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:44]: A lot of these variables are only needed inside `beforeEach` and they should be moved there.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 9 above)
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:364]: I believe this was a change sonarlint wanted - I'll change it back and see if it objects!

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 10 above)
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:368]: This was another sonarlint one - I wrote `!status` initially and it showed `Avoid using boxed "Boolean" types directly in boolean expressions`

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java [hunk 15]
```diff
@@ -0,0 +1,423 @@
...
+        expect(userManager.updateTeacherPendingFlag(3L, false)).andReturn(prepareTestUser(3L, false));
+
+        expect(emailManager.getEmailTemplateDTO("teacher_declined")).andThrow(new ContentManagerException("Content not found"));
+
+        expect(userManager.getUserDTOById(2L)).andReturn(prepareTestUser(2L, true));
+        expect(userManager.updateTeacherPendingFlag(2L, false)).andReturn(prepareTestUser(2L, false));
+
+        expect(emailManager.getEmailTemplateDTO("teacher_declined")).andReturn(new EmailTemplateDTO());
+        emailManager.sendTemplatedEmailToUser(anyObject(RegisteredUserDTO.class), anyObject(EmailTemplateDTO.class),
+            anyObject(Map.class), eq(EmailType.SYSTEM));
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:221]: `anyObject(Map.class)` will display a warning about unchecked assignments, but simply calling `anyObject()` instead should infer the right type (I think).

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+        userPreferenceManager, eventBookingManager, segueJobService, externalAccountManager, misuseMonitor, emailManager);
+  }
+
+  @Nested
+  class ModifyUsersTeacherPendingStatus {
+    @Nested
+    class FailureResponses {
+      @Test
+      void modifyUsersTeacherPendingStatus_anonymousUser_returnsUnauthorised() throws NoUserLoggedInException {
+        HttpServletRequest mockRequest = replayMockServletRequest();
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:82]: You seem to be calling this before each test, so it sounds like a good candidate to be moved to `beforeEach`.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java [hunk 17]
```diff
@@ -0,0 +1,423 @@
...
+              response.readEntity(SegueErrorResponse.class).getErrorMessage());
+        }
+
+        verify(userManager);
+      }
+
+      @Test
+      void modifyUsersTeacherPendingStatus_nonAdminUser_returnsForbidden() throws NoUserLoggedInException {
+        HttpServletRequest mockRequest = replayMockServletRequest();
+        RegisteredUserDTO currentUser = new RegisteredUserDTO();
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:101]: It sounds like in a lot of these tests you could have simply declared the `currentUser` as a mock, since you don't really need to interact with it at all, and you don't care about its contents, nor any of its methods need to return anything in particular. It's not a big deal in this case, since it's a DTO, so there is no business logic inside anyway, but still something to think about. Making it a mock would also signal more clearly that you don't really care about this user, and it's only here because it's needed as a parameter of other functions.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java [hunk 18]
```diff
@@ -0,0 +1,423 @@
...
+        expect(userManager.getCurrentRegisteredUser(mockRequest)).andReturn(currentUser);
+        expect(isUserAnAdminOrEventManager(userManager, currentUser)).andReturn(true);
+
+        expect(userManager.getUserDTOById(2L)).andThrow(new NoUserException("No user found with this ID!"));
+
+        expect(userManager.getUserDTOById(3L)).andReturn(prepareTestUser(3L, true));
+        expect(userManager.updateTeacherPendingFlag(3L, false)).andReturn(prepareTestUser(3L, false));
+
+        expect(emailManager.getEmailTemplateDTO("teacher_declined")).andReturn(new EmailTemplateDTO());
+        emailManager.sendTemplatedEmailToUser(anyObject(RegisteredUserDTO.class), anyObject(EmailTemplateDTO.class),
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:157]: Could you explain how this work, please? I'm not familiar with `expectLastCall`. Is the reason why this line is not inside `expect` because it comes right before `expectLastCall`?

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+          assertEquals("One or more users could not be found: [2]. ",
+              response.readEntity(SegueErrorResponse.class).getErrorMessage());
+        }
+
+        verify(userManager);
+        verify(emailManager);
+      }
+
+      @Test
+      void modifyUsersTeacherPendingStatus_emailSendFailureForOneTarget_returnsBadRequest()
```
Comment (jmgq) [inline src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:175]: As mentioned in my other comment, it sounds like this should be a server error instead.

Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:175]: On second thought, how important is this, actually? Maybe this is acceptable? This is more of a business logic question: if we fail to send some e-mails, what does that mean for the end user? Should we be returning a 200 response, but with a warning message or something?

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 10 above)
Comment (jmgq) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:368]: I think that the right solution depends on whether it's acceptable to pass `null` or not. If the only two right values are either `true` or `false`, then you should use the primitive type (`boolean`, rather than `Boolean`), as that one cannot be `null`.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 9 above)
Comment (jmgq) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:364]: I think it may have complained before if you were using string concatenation in the `log.info` string, but you don't seem to be doing that now, so it should be fine.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java (hunk 19 above)
Comment (hannah-whelan) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:175]: If 200 response comes back that's "ok" so no message comes back to the end user on the front end (and they aren't inspecting the network response). 

If it returns an error code under 500 it doesn't totally bomb out the page, it shows an error popup with extra information. So I was trying to do this in a user friendly way so that a user might get a popup that notifies of email send failures (for whatever reason) but the teacher_pending status of those users in the DB would still have been modified. In which case the user (who is a STEM staff member) can manually send an email. 

If we returned a 500 error it would takeover the page, and be unclear to the user that it had updated the statuses and just failed at the email send point.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java (hunk 18 above)
Comment (hannah-whelan) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:157]: Robert explained to me that EasyMock expect() can’t be used to mock void methods and that I should use expectLastCall() instead

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java (hunk 19 above)
Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:175]: Thanks for explaining, Hannah. That makes sense, but you're seeing all this from the perspective of the current implementation of the frontend, however when designing the API you need to make sure that it adheres to the conventions that RESTful APIs normally follow. For instance, the server failing to send some emails is not the client's fault, so you cannot return a 4xx error. It should be a 500 error, since it's a server error. However, if sending these emails are not completely necessary, one could argue that if the server manages to update all users' teacher pending status, but it failed to send some emails, then that could be considered a 200 response, because the main task (updating the status) was successful. In the body of that response, you could have a JSON with an optional attribute called `warning`, or something like that, and populate that with the IDs of the users that you didn't manage to send the emails to, for instance. Does that make sense?

Comment (hannah-whelan) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java]: Changed this now so that email send failure sends back 200 response with a message (and front end adjusted to display that message in a popup).

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 20]
```diff
@@ -281,6 +294,112 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+                                                               final List<Long> userIds) {
+
+    try {
+      RegisteredUserDTO requestingUser = userManager.getCurrentRegisteredUser(request);
+      if (!isUserAnAdminOrEventManager(userManager, requestingUser)) {
+        return new SegueErrorResponse(Status.FORBIDDEN, ACCESS_DENIED_MESSAGE)
+            .toResponse();
+      }
+
+      Map<String, List<Long>> failedUpdates = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java]: If I gave each user their own error (if any happened), I would still need to check if a specific error had been recorded surely? And gather all the userIDs with that error. That's why I felt like it made the most sense to record the 2 errors I'm interested in (email send failure and no user found) and store a list of userIds that I can then grab for the relevant response message

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 21]
```diff
@@ -281,6 +294,114 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+  @Path("/users/change_teacher_pending/{status}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Operation(summary = "Update teacher_pending to true or false for a list of possible user IDs.",
+      description = "This endpoint requires an admin user to be logged in. If updating teacher_pending to false it "
+          + "also sends an email to decline teacher account upgrade. If email send fails, response is still OK but "
+          + "with added message to notify user")
+  public synchronized Response modifyUsersTeacherPendingStatus(@Context final HttpServletRequest request,
+                                                               @PathParam("status") final Boolean status,
+                                                               final List<Long> userIds) {
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:315]: Have you (manually) tested what happens when `userIds`:

- Is not set.
- It's set to null.
- It contains non-numeric values (e.g. one of the items of the array is a string, for instance).
- It's not a list.

If any of those scenarios returns a 500 error, it means that we need to account for them, and return a 400 error instead. It is possible that the framework already manages a lot of these (if not all) for you, but it would be good to know for sure.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 22]
```diff
@@ -281,6 +294,114 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+  @POST
+  @Path("/users/change_teacher_pending/{status}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Operation(summary = "Update teacher_pending to true or false for a list of possible user IDs.",
+      description = "This endpoint requires an admin user to be logged in. If updating teacher_pending to false it "
+          + "also sends an email to decline teacher account upgrade. If email send fails, response is still OK but "
+          + "with added message to notify user")
+  public synchronized Response modifyUsersTeacherPendingStatus(@Context final HttpServletRequest request,
+                                                               @PathParam("status") final Boolean status,
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:314]: Does this one have to be `Boolean` or can it be `boolean`? You've changed it to `boolean` downstream, which suggests that `null` values are not allowed. If that's the case, I would change this one to `boolean`, unless the framework requires it to be `Boolean`.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java (hunk 15 above)
Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java]: I should have mentioned this, but this happens in a couple other places. Does your IDE highlight it? Given that this is all new code, it would be better to ensure it has no Warnings.

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java (hunk 18 above)
Comment (jmgq) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java:146]: Thank you, that makes sense!

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 23]
```diff
@@ -281,6 +294,114 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+                                                               final List<Long> userIds) {
+
+    Map<String, List<Long>> failedUpdates = new HashMap<>();
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:317]: I think this should probably be a `Set` instead of a `List`, as you don't care about order, and you don't want repeated IDs for each type of error anyway.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 24]
```diff
@@ -281,6 +294,114 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
...
+    if (failedUpdates.containsKey(USERS_NOT_FOUND)) {
+      String errorMessage =
+          String.format("One or more users could not be found: %s", failedUpdates.get(USERS_NOT_FOUND));
+      if (failedUpdates.containsKey(FAILED_TO_SEND)) {
+        errorMessage += String.format(" Emails could not be sent to userIds: %s", failedUpdates.get(FAILED_TO_SEND));
+      }
+      return new SegueErrorResponse(Status.BAD_REQUEST, errorMessage).toResponse();
+    }
+
+    String responseString = "Teacher pending status updated to " + status + " for requested userIds: " + userIds;
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:350]: Small one, but it looks like this could be part of an if/else, since the response is always going to be either this one or the one below. The end result is the same, but by making it part of an if/else, it's clearer. Also, `responseMessage` is a bit more meaningful.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (jmgq) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java]: >I would still need to check if a specific error had been recorded surely?

Yeah, but you can still do that with the way I suggested.

If you want to keep this like this and also prefer it to extracting it to its own class, that's fine, but you should change some things, because it really is too noisy and distracting. For instance, have a look at this code:

```java
      if (!failedUpdates.containsKey(FAILED_TO_SEND)) {
        List<Long> userList = new ArrayList<>();
        userList.add(userId);
        failedUpdates.put(FAILED_TO_SEND, userList);
      } else {
        List<Long> userList = failedUpdates.get(FAILED_TO_SEND);
        userList.add(userId);
      }
```

This is not only duplicated code, but it's 8 lines of code when all you really want is to record that a certain error happened when dealing with a certain user.

One way to make this easier to read is to initialise both `USERS_NOT_FOUND` and `FAILED_TO_SEND` to empty lists (or probably sets). For instance:

```java
    Map<String, Set<Long>> failedUpdates = new HashMap<>();
    failedUpdates.put(USERS_NOT_FOUND, new HashSet<>());
    failedUpdates.put(FAILED_TO_SEND, new HashSet<>());
```

Now you know that both keys are always present and initialised. Then those 8 lines simply become:

```java
failedUpdates.get(FAILED_TO_SEND).add(userId);
```

Of course, it also means that `failedUpdates.containsKey(FAILED_TO_SEND)` becomes `!failedUpdates.get(FAILED_TO_SEND).isEmpty()`, but that's not a big deal.

[the remark this answers is not in the document]
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java]: Ok yeah that's much better

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java (hunk 21 above)
Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:315]: userIds:

Is not set = 500 
It's set to null = 500
It contains non-numeric values= 400 bad request
It's not a list = 500

The front end requires userIds to be `number[]` so are we envisaging someone using this endpoint separately from the front end? Or just being doubly sure in case a future dev decides to make changes to the front end and breaks things?

File: src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java (hunk 15 above)
Comment (hannah-whelan) [reply src/test/java/uk/ac/cam/cl/dtg/segue/api/AdminFacadeTest.java]: Yeah it was highlighting, not sure why I missed them! They should be sorted now

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 25]
```diff
@@ -63,10 +63,14 @@
 import java.io.IOException;
 import java.security.InvalidKeyException;
 import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
```
Comment (jmgq) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:66]: This one is no longer needed.

Comment (hannah-whelan) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java]: It did highlight it in my IDE at least, I just missed it 😬

Comment (jmgq) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java]: You can tell IntelliJ to automatically optimise the imports on save. I have not been able to make it work so that it also complies with the import rules set by CheckStyle, but Robert has, so he may be able to help you.

**Review by jmgq — APPROVED** [review]
